packages/data-types/src/datatype.ts

```diff
@@ -0,0 +1,151 @@
+        });
+
+        return `
+            ${baseTypeList.join('\n')}
+            ${[...new Set(customTypesList)].join('\n')}
+        `;
+    }
+
+    constructor({ version, fields:typesConfig }: DataTypeConfig, typeName?: string) {
+        if (version == null) throw new ts.TSError('No version was specified in type_config');
+        const typeManager = new TypesManager(version);
+        const types:BaseType[] = [];
+        for (const key in typesConfig) {
+            // ignore metadata settings
+            if (key.charAt(0) !== '$') {
+                const typeConfig = typesConfig[key];
+                types.push(typeManager.getType(key, typeConfig));
+            }
+        }
+        if (typeName != null) this._name = typeName;
+        this._types = types;
+    }
+
+    toESMapping({ typeName, settings: settingsConfig, mappingMetaData }: MappingConfiguration) {
+        const argAnalyzer = ts.get(settingsConfig || {}, ['analysis', 'analyzer'], {});
+        const argTokenizer = ts.get(settingsConfig || {}, ['analysis', 'tokenizer'], {});
+
+        const analyzer = { ...argAnalyzer };
+        const tokenizer = { ...argTokenizer };
+
+        const properties = this._types.reduce((accum, type) => {
+            const { mapping, analyzer: typeAnalyzer = {}, tokenizer: typeTokenizer = {} } = type.toESMapping();
+
+            // get mapping configuration
+            for (const key in mapping) {
+                accum[key] = mapping[key];
+            }
+
+            // get analyzer configuration
+            for (const key in typeAnalyzer) {
+                analyzer[key] = typeAnalyzer[key];
+            }
+
+             // get tokenizer configuration
+            for (const key in typeTokenizer) {
+                tokenizer[key] = typeTokenizer[key];
+            }
+
+            return accum;
+        }, {});
+
+        const analysis = {
+            analysis: {
+                analyzer,
+                tokenizer
+            }
+        };
+
+        // TODO: what default settings and analyzers should go here?
+        const settings: EsMapSettings = {
+            ...settingsConfig,
+            ...analysis
+        };
+
+        const mappingConfig = {
+            properties
+        };
+
+        if (mappingMetaData != null && ts.isPlainObject(mappingMetaData)) {
+            for (const key in mappingMetaData) {
+                mappingConfig[key] = mappingMetaData[key];
+            }
+        }
+
+        return {
+            mappings: {
+                [typeName]: mappingConfig
+            },
+            settings
+        };
+    }
+
+    toGraphQl(typeName?: string|null|undefined, typeInjection?:string) {
+        const { baseType, customTypes } = this.toGraphQLTypes(typeName, typeInjection);
+
+        return  `
+            ${baseType}
+            ${[...new Set(customTypes)].join('\n')}
+        `;
+    }
+
+    toGraphQLTypes(typeName?: string|null|undefined, typeInjection?:string) {
```

Comment:
This can simplified.

```suggestion
    toGraphQLTypes(typeName = this._name, typeInjection?:string) {
```